client/service/src/lib.rs

```diff
@@ -342,6 +364,29 @@ async fn build_network_future<
 			state_sink = status_sinks.state.next().fuse() => {
 				state_sink.send(network.network_state());
 			}
+
+			_ = (&mut handle_bitswap_events).fuse() => {}
+		}
+	}
+}
+
+/// Handle a `BitswapEvent`, reading and writing from the `OffchainStorage`.
+pub async fn handle_bitswap_event<B: BlockT, H: sc_network::ExHashT, S: IpldStore<Params=IpldStoreParams>>(
+	service: &Arc<sc_network::NetworkService<B, H>>,
+	event: sc_network::BitswapEvent,
+	storage: S,
+) -> Result<(), libipld::error::Error> {
+	match event {
+		sc_network::BitswapEvent::ReceivedBlock(_, cid, data) => {
+			storage.insert(libipld::Block::new(cid.clone(), data.to_vec())?).await?;
```

Comment:
ipfs-embed moved away from get making any changes to the stored blocks. in ipfs-embed this will actually not insert anything at all as the block isn't pinned. I see the need for recording the timestamps and having a mixture of pins, refcounts and timestamps determine if a block is live or not. That way the store can be used as a cache without leading to storage leakage.